### ISSUE TITLE
panic: deadlock detected when coming back from sleep

### ISSUE THREAD

**Issue by fbreve** [syncthing/syncthing#1536]
Syncthing is eventually crashing in my Windows 8.1 system. Here is part of the last crash log file:
```
[ZVTWH] 00:20:40 INFO: Puller (folder "Doutorado", dir "Simulações\\Matlab\\Daniel\\Pacote2\\Reciprocal-kNN-Distance\\Linear-k30"): delete: remove d:\Users\Fabricio\Documents\Doutorado\Simulações\Matlab\Daniel\Pacote2\Reciprocal-kNN-Distance\Linear-k30: A pasta não está vazia.
[ZVTWH] 00:20:40 INFO: Puller (folder "Doutorado", dir "Simulações\\Matlab\\Daniel\\Pacote2\\OriginalInput"): delete: remove d:\Users\Fabricio\Documents\Doutorado\Simulações\Matlab\Daniel\Pacote2\OriginalInput: A pasta não está vazia.
[ZVTWH] 00:20:46 INFO: Puller (folder "Doutorado", dir "Simulações\\Matlab\\Daniel\\Pacote2\\Reciprocal-kNN-Distance\\Linear-k30"): delete: remove d:\Users\Fabricio\Documents\Doutorado\Simulações\Matlab\Daniel\Pacote2\Reciprocal-kNN-Distance\Linear-k30: A pasta não está vazia.
[ZVTWH] 00:20:46 INFO: Puller (folder "Doutorado", dir "Simulações\\Matlab\\Daniel\\Pacote2\\OriginalInput"): delete: remove d:\Users\Fabricio\Documents\Doutorado\Simulações\Matlab\Daniel\Pacote2\OriginalInput: A pasta não está vazia.
[ZVTWH] 00:22:13 DEBUG: rwFolder/Fotos@0xc082246580 adjusting curVer 5102479
[ZVTWH] 00:22:17 INFO: Puller (folder "Doutorado", dir "Simulações\\Matlab\\Daniel\\Pacote2\\Reciprocal-kNN-Distance\\Linear-k30"): delete: remove d:\Users\Fabricio\Documents\Doutorado\Simulações\Matlab\Daniel\Pacote2\Reciprocal-kNN-Distance\Linear-k30: A pasta não está vazia.
[ZVTWH] 00:22:17 INFO: Puller (folder "Doutorado", dir "Simulações\\Matlab\\Daniel\\Pacote2\\OriginalInput"): delete: remove d:\Users\Fabricio\Documents\Doutorado\Simulações\Matlab\Daniel\Pacote2\OriginalInput: A pasta não está vazia.
[ZVTWH] 00:22:33 INFO: Puller (folder "Doutorado", dir "Simulações\\Matlab\\Daniel\\Pacote2\\Reciprocal-kNN-Distance\\Linear-k30"): delete: remove d:\Users\Fabricio\Documents\Doutorado\Simulações\Matlab\Daniel\Pacote2\Reciprocal-kNN-Distance\Linear-k30: A pasta não está vazia.
[ZVTWH] 00:22:33 INFO: Puller (folder "Doutorado", dir "Simulações\\Matlab\\Daniel\\Pacote2\\OriginalInput"): delete: remove d:\Users\Fabricio\Documents\Doutorado\Simulações\Matlab\Daniel\Pacote2\OriginalInput: A pasta não está vazia.
[ZVTWH] 00:22:33 WARNING: Folder "Doutorado" isn't making progress - check logs for possible root cause. Pausing puller for 1m0s.
[ZVTWH] 02:46:26 INFO: Paused state detected, possibly woke up from standby. Restarting in 1m0s.
[ZVTWH] 02:46:26 INFO: Connection to [REMOVED] closed: ping timeout
[ZVTWH] 02:46:32 INFO: Connection to [REMOVED] closed: WSARecv tcp 192.168.2.35:29155: Foi forçado o cancelamento de uma conexão existente pelo host remoto.
panic: deadlock detected
goroutine 50 [running]:
github.com/syncthing/syncthing/internal/model.func·029()
	/go/src/github.com/syncthing/syncthing/internal/model/util.go:32 +0x221
created by github.com/syncthing/syncthing/internal/model.deadlockDetect
	/go/src/github.com/syncthing/syncthing/internal/model/util.go:35 +0xdd
goroutine 1 [chan receive, 2688 minutes]:
main.syncthingMain()
	/go/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:631 +0x28ae
main.main()
	/go/src/github.com/syncthing/syncthing/cmd/syncthing/main.go:362 +0x2274
goroutine 5 [syscall, 2690 minutes]:
os/signal.loop()
	c:/go/src/os/signal/signal_unix.go:21 +0x26
created by os/signal.init·1
	c:/go/src/os/signal/signal_unix.go:27 +0x3c
goroutine 7 [select]:
github.com/syncthing/syncthing/internal/events.(*Subscription).Poll(0xc0820072a0, 0xdf8475800, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/syncthing/syncthing/internal/events/events.go:173 +0x3c6
github.com/syncthing/syncthing/internal/events.(*BufferedSubscription).pollingLoop(0xc082008580)
	/go/src/github.com/syncthing/syncthing/internal/events/events.go:209 +0x4f
created by github.com/syncthing/syncthing/internal/events.NewBufferedSubscription
	/go/src/github.com/syncthing/syncthing/internal/events/events.go:203 +0x1cb
goroutine 8 [chan receive]:
main.trackCPUUsage()
	/go/src/github.com/syncthing/syncthing/cmd/syncthing/gui_windows.go:37 +0x496
created by main.init·3
	/go/src/github.com/syncthing/syncthing/cmd/syncthing/gui_windows.go:17 +0x2c
goroutine 9 [chan receive, 2690 minutes]:
github.com/syncthing/syncthing/internal/config.(*Wrapper).Serve(0xc0820fd200)
	/go/src/github.com/syncthing/syncthing/internal/config/wrapper.go:79 +0x7c
created by github.com/syncthing/syncthing/internal/config.Wrap
	/go/src/github.com/syncthing/syncthing/internal/config/wrapper.go:54 +0x13a
goroutine 10 [select]:
github.com/syndtr/goleveldb/leveldb/util.(*BufferPool).drain(0xc0820862a0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/util/buffer_pool.go:205 +0x225
created by github.com/syndtr/goleveldb/leveldb/util.NewBufferPool
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/util/buffer_pool.go:236 +0x253
goroutine 11 [select]:
github.com/syndtr/goleveldb/leveldb.(*DB).compactionError(0xc0820802c0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:120 +0x493
created by github.com/syndtr/goleveldb/leveldb.openDB
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db.go:126 +0x850
goroutine 12 [select]:
github.com/syndtr/goleveldb/leveldb.(*DB).mpoolDrain(0xc0820802c0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_state.go:73 +0x151
created by github.com/syndtr/goleveldb/leveldb.openDB
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db.go:127 +0x86a
goroutine 13 [runnable, locked to thread]:
syscall.FlushFileBuffers(0x16d4, 0x0, 0x0)
	c:/go/src/syscall/zsyscall_windows.go:919 +0x76
syscall.Fsync(0x16d4, 0x0, 0x0)
	c:/go/src/syscall/syscall_windows.go:502 +0x41
os.(*File).Sync(0xc0bc0b5060, 0x0, 0x0)
	c:/go/src/os/file_posix.go:129 +0x71
github.com/syndtr/goleveldb/leveldb/storage.fileWrap.Sync(0xc0bc0b5060, 0xc08f07f640, 0x0, 0x0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/storage/file_storage.go:347 +0x41
github.com/syndtr/goleveldb/leveldb/storage.(*fileWrap).Sync(0xc0d409b330, 0x0, 0x0)
	<autogenerated>:49 +0xb8
github.com/syndtr/goleveldb/leveldb.(*tWriter).finish(0xc0e24a5c20, 0x0, 0x0, 0x0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/table.go:504 +0x10d
github.com/syndtr/goleveldb/leveldb.(*tableCompactionBuilder).flush(0xc0881d05a0, 0x0, 0x0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:420 +0x5c
github.com/syndtr/goleveldb/leveldb.(*tableCompactionBuilder).run(0xc0881d05a0, 0xc0e10bf8e8, 0x0, 0x0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:481 +0x98d
github.com/syndtr/goleveldb/leveldb.(*DB).compactionTransact(0xc0820802c0, 0xa680f0, 0xb, 0x2d68a8, 0xc0881d05a0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:215 +0x27e
github.com/syndtr/goleveldb/leveldb.(*DB).tableCompaction(0xc0820802c0, 0xc090c66d00, 0x0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:596 +0x10db
github.com/syndtr/goleveldb/leveldb.(*DB).tableAutoCompaction(0xc0820802c0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:641 +0x57
github.com/syndtr/goleveldb/leveldb.(*DB).tCompaction(0xc0820802c0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:833 +0x3a7
created by github.com/syndtr/goleveldb/leveldb.openDB
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db.go:130 +0x8ae
goroutine 14 [select, 142 minutes]:
github.com/syndtr/goleveldb/leveldb.(*DB).mCompaction(0xc0820802c0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_compaction.go:759 +0x28a
created by github.com/syndtr/goleveldb/leveldb.openDB
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db.go:131 +0x8c8
goroutine 15 [select, 144 minutes]:
github.com/syndtr/goleveldb/leveldb.(*DB).jWriter(0xc0820802c0)
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db_write.go:37 +0x19e
created by github.com/syndtr/goleveldb/leveldb.openDB
	/go/src/github.com/syncthing/syncthing/Godeps/_workspace/src/github.com/syndtr/goleveldb/leveldb/db.go:132 +0x8e2
goroutine 96720 [sleep]:
github.com/syncthing/syncthing/internal/model.sendIndexes(0x3339e50, 0xc084ff63c0, 0xc0820f90f0, 0x6, 0xc082648cf0, 0xc082728690)
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:926 +0x343
created by github.com/syncthing/syncthing/internal/model.(*Model).AddConnection
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:873 +0x580
goroutine 254 [semacquire, 144 minutes]:
sync.(*Mutex).Lock(0xc0835db130)
	c:/go/src/sync/mutex.go:66 +0xda
sync.(*RWMutex).Lock(0xc0835db130)
	c:/go/src/sync/rwmutex.go:82 +0x3a
github.com/syncthing/syncthing/internal/model.(*Model).ScanFolderSub(0xc0835db0a0, 0xc0820f8630, 0x7, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:1102 +0x3dd
github.com/syncthing/syncthing/internal/model.(*Model).ScanFolder(0xc0835db0a0, 0xc0820f8630, 0x7, 0x0, 0x0)
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:1093 +0x69
github.com/syncthing/syncthing/internal/model.(*rwFolder).Serve(0xc0821393f0)
	/go/src/github.com/syncthing/syncthing/internal/model/rwfolder.go:226 +0x1744
created by github.com/syncthing/syncthing/internal/model.(*Model).StartFolderRW
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:162 +0x62d
goroutine 49 [select, 784 minutes]:
github.com/syncthing/syncthing/internal/model.(*ProgressEmitter).Serve(0xc08343a420)
	/go/src/github.com/syncthing/syncthing/internal/model/progressemitter.go:48 +0x8bf
created by github.com/syncthing/syncthing/internal/model.NewModel
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:117 +0x613
goroutine 51 [chan receive]:
github.com/syncthing/syncthing/internal/model.func·029()
	/go/src/github.com/syncthing/syncthing/internal/model/util.go:31 +0x1c4
created by github.com/syncthing/syncthing/internal/model.deadlockDetect
	/go/src/github.com/syncthing/syncthing/internal/model/util.go:35 +0xdd
goroutine 52 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082922320, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 53 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc0822c7ef0, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 54 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc08218e4b0, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 55 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc08218e7d0, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 56 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082922e10, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 34 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082922fa0, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 35 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082923c20, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 57 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082746230, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 58 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc08270e690, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 59 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082729630, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 60 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082418f50, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 146 [semacquire, 144 minutes]:
sync.(*RWMutex).RLock(0xc0835db160)
	c:/go/src/sync/rwmutex.go:36 +0x66
github.com/syncthing/syncthing/internal/model.(*Model).ConnectedTo(0xc0835db0a0, 0xb710ce21655bfe7f, 0x13c4970f90c86307, 0x28e1938100d90f79, 0x5cfd0f6f4c8a7314, 0x0)
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:770 +0x42
main.dialTLS(0xc0835db0a0, 0xc083472120, 0xc0820f6b40)
	/go/src/github.com/syncthing/syncthing/cmd/syncthing/connections.go:193 +0x214
created by main.listenConnect
	/go/src/github.com/syncthing/syncthing/cmd/syncthing/connections.go:31 +0x1a9
goroutine 62 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc0827281e0, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 253 [chan receive]:
github.com/syncthing/syncthing/internal/versioner.func·002()
	/go/src/github.com/syncthing/syncthing/internal/versioner/staggered.go:113 +0xbe
created by github.com/syncthing/syncthing/internal/versioner.NewStaggered
	/go/src/github.com/syncthing/syncthing/internal/versioner/staggered.go:116 +0x6af
goroutine 126374 [sleep]:
github.com/syncthing/syncthing/internal/model.sendIndexes(0x3339e50, 0xc08a53e9e0, 0xc08205e960, 0x11, 0xc08238a660, 0xc0822c7ef0)
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:926 +0x343
created by github.com/syncthing/syncthing/internal/model.(*Model).AddConnection
	/go/src/github.com/syncthing/syncthing/internal/model/model.go:873 +0x580
goroutine 65 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082484e60, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 66 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082747900, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 67 [select]:
github.com/syncthing/syncthing/internal/ignore.(*Matcher).clean(0xc082728690, 0x68c61714000)
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:153 +0x1c0
created by github.com/syncthing/syncthing/internal/ignore.New
	/go/src/github.com/syncthing/syncthing/internal/ignore/ignore.go:53 +0xcd
goroutine 68 [select]:
```

**Comment by AudriusButkevicius** [comment]
To me this looks like your PC went to sleep in some state which caused us to panic, hence marking as a bug.

**Comment by calmh** [comment]
I think the deadlock detection doesn't handle sleep very well. We start two goroutines that race to completion - one that grabs and releases the lock and one that sleeps for an amount of time. If the lock takes two seconds to acquire and the computer goes to sleep for two hours after one second, then when waking up it will seem like we've tried to acquire the lock for two hours and we'll panic.

Probably this should be retired. It was useful for a while when we had deadlock bugs in the model, but lately I've only seen this triggered by hardware problems and now this...

**Comment by AudriusButkevicius** [comment]
Well it pointed out dst deadlocks, etc, hence is useful, but perhaps should only run for prerelease or debug/dev builds.

**Comment by fbreve** [comment]
I forgot to mention. It's probably not related, but this log also shows some "Info: Puller" problems. I uncompressed some attachment I got by email into a folder. I realized it was the wrong attachment, so I deleted the folder and uncompressed the right one (same folder name). Now this error seems to live forever.

**Comment by calmh** [comment]
@fbreve "The folder is not empty" (thanks google translate :) is because syncthing is supposed to remove a directory that is supposed to be empty, but isn't. Can you check what the files in that directory are? They could files covered by an ignore pattern (in which case you should fix that) or they could be temporary files left over from a failed sync (in which case *we* should fix *that*)...

**Comment by fbreve** [comment]
The files there were deleted in the source machine together with the directory.
They are not files covered by an ignore pattern. They are not temporary files.
They should have been excluded with the directory.

**Comment by calmh** [comment]
Yes. But something is present on the *destination* device to block the removal of the directories - check what that is?

**Comment by fbreve** [comment]
I couldn't find anything. There's nothing ignored. Permissions are ok. There are no open files. I was able to remove them manually just fine.

**Comment by calmh** [comment]
Right. But syncthing tried to remove the directories and got "directory not empty" back from the OS, so there was something in them. If you deleted them from the Finder or Explorer, it deleted the directories and any files in them. If there was nothing visible I'll assume they were temp files left over for some reason...

**Comment by fbreve** [comment]
The files were there, but they were removed in the source system, so they should have been removed in the destination system as well. Maybe it has something to do with the fact that I created and deleted them in a short period?

There weren't any hidden or temporary files in that directory, Explorer is set to show hidden files by default here.

Anyway, I'll let you know if it happens again or if I can reproduce it.

Thank you.
